lib/Doctrine/ORM/Tools/Pagination/Paginator.php

```diff
@@ -150,14 +155,21 @@ public function getIterator()
 
             $subQuery->setFirstResult($offset)->setMaxResults($length);
 
-            $ids = array_map('current', $subQuery->getScalarResult());
+            $foundIdRows = $subQuery->getScalarResult();
 
-            $whereInQuery = $this->cloneQuery($this->query);
             // don't do this for an empty id array
-            if (count($ids) === 0) {
+            if ($foundIdRows === []) {
                 return new \ArrayIterator([]);
             }
 
+            $whereInQuery = $this->cloneQuery($this->query);
+            $em           = $subQuery->getEntityManager();
+            $connection   = $em->getConnection();
+            $idType       = $this->getIdentifiersQueryScalarResultType($subQuery, $em);
```

Comment:
Wouldn't we keep things a bit more contained if we perform this conversion in `WhereInWalker`? There's a lot going on there too but we already have the class metadata and it might be easier to detect the type?

Reply:
And I think we would also mitigate possible performance impact due to double query parsing (although query cache reduces that)

Reply:
> Wouldn't we keep things a bit more contained if we perform this conversion in `WhereInWalker`? 

Initially attempted this, but `WhereInWalker` does not have access to the DQL parameters, only to the DQL string/AST.